src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,176 @@
-""" This is a dummy example """
+import requests
+import pandas as pd
+from pandas import DataFrame
+import plotly.express as px  # type: ignore
+import streamlit as st
+import time
+from typing import Any, Dict
+
+API_URL = "https://archive-api.open-meteo.com/v1/archive"
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+VARIABLES = ["temperature_2m", "precipitation", "wind_speed_10m"]
+
+
+PLOT_INFO = {
+    "temperature_2m": {"color": "red", "y_label": "Mean Temperature (ºC)"},
+    "precipitation": {"color": "blue", "y_label": "Accumulated precipitation (mm)"},
+    "wind_speed_10m": {"color": "green", "y_label": "Max. Wind speed 10 meters (km/h)"},
+}
+
+
+def get_data_meteo_api(city: str, magnitude: str) -> Dict[str, Any]:
+    params = {
+        "latitude": str(COORDINATES[city]["latitude"]),
+        "longitude": str(COORDINATES[city]["longitude"]),
+        "start_date": "2010-01-01",
+        "end_date": "2019-12-31",
+        "hourly": magnitude,
+    }
+
+    r = requests.get(API_URL, params=params)
+    print(r.status_code)
+    data = r.json()
+    return data
+
+
+# --------------------------------------------
+# Generic function for APIs conection
+# Not used in the main()
+# -------------------------------------------
+def make_api_request(url, params=None, headers=None, retries=3, cooldown=10):
+    attempt = 0
+    while attempt < retries:
+        try:
+            response = requests.get(url, params=params, headers=headers)
+            status_code = response.status_code
+
+            if status_code == 200:
+                return response.json()
+
+            elif status_code == 429:  # Rate limit alcanzado
+                print(f"Rate limit alcanzado. Esperando {cooldown} segundos...")
+                time.sleep(cooldown)
+
+            elif 500 <= status_code < 600:  # Error en el servidor
+                print(f"Error en el servidor ({status_code}). Reintentando...")
+                attempt += 1
+                time.sleep(cooldown)
+
+            else:
+                response.raise_for_status()
+
+        except requests.exceptions.RequestException as e:
+            print(f"Error al realizar la solicitud: {e}")
+            attempt += 1
+            time.sleep(cooldown)
+
+    raise Exception(f"La solicitud a la API falló después de {retries} intentos.")
+
+
+# -----------------------
+# Params: data = API schema
+# Return: data_proc = DataFrame[time,temperature_2m]
+# ----------------------
+def process_data(data: Dict[str, Any], magnitude: str) -> DataFrame:
```

Comment:
`process_data` is too vague a name. Let's put some thought on giving concrete and descriptive names to our functions so the code become as self-documented as possible.